### DISCUSSION TITLE
Python module running on pyodide interacting directly with the DOM within a notebook cell.

### DISCUSSION THREAD

**Discussion by sspickle** [jupyterlite/jupyterlite#747] [Q&A] [answered]
I've been reading about this project, which looks great! I'd like to create a python module that adds a canvas element to the DOM, but rather than go through the traditional comm channel architecture of ipywdigets, since it's using pyodide, it should be able to access the canvas object/context/etc. directly, no? I've already got a proof of concept that runs in a web page and it works well, but I'm not sure how to integrate the module into jupyterlite. Is this crazy? Thank you for any guidance.

**Comment by jtpio** [comment]
Thanks @sspickle.

> I'd like to create a python module that adds a canvas element to the DOM

Maybe `ipycanvas` would actually be a good solution for your use case? https://github.com/martinRenou/ipycanvas

It works in JupyterLite and lets you draw on an HTML canvas in Python. It is a Jupyter Widget but the high level API hides the widgets comm channel implementation details. You can check the `ipycanvas` demo with JupyterLite on the documentation website: https://ipycanvas.readthedocs.io/en/latest/lite/lab/

**Reply by sspickle** [reply]
Let me explain a bit more. I'm working on a version of WebVPython that's built on pyodide. I have a rough proof of concept here:

[SHO_Example.py](https://glowscriptdev.spvi.net/?docurl=https%3A%2F%2Fstorage.googleapis.com%2Fuindysjs%2FSHO_Example.py)

There's a python proxy class that represents all the various javascript objects that are used by the renderer to render the scene. Any time you modify the python object it updates the corresponding javascript object directly. I'm hoping there's a way to use this implementation with little change in jupyterlite rather than serializing all the changes and sending them over the comms. Does it seem possible?

thanks,
-steve

**Comment by bollwyvl** [comment, marked as answer]
Current contributions have been concentrating on presenting as close an analog to the traditional jupyter client/server stack as possible, and emulating that experience.

In the `pyolite` shipped with jupyterlite, pyodide is running inside its emscripten sandbox inside a WebWorker, so cannot access the DOM, and a number of other `window`-only APIs. The intent of this is to keep the kernel from wedging the main UI... and could in the future be used to restrict kernels to entirely separate domains, which would improve the security posture... to a limited extent.

Alternative kernels are possible, such as the [xeus-python-kernel](https://github.com/jupyterlite/xeus-python-kernel)... but it also uses a webworker, for much the same reasons.

The javascript kernel even isolates its scope to a hidden iframe.

So: one could likely further proxy this with comlink (or some other technique), but I don't believe anyone has investigated this.
